File: samples-java/datalayer.provider.simple/app/src/main/java/datalayer/provider/simple/ProviderNodeCallbackTimestamp.java
/*
 * SPDX-FileCopyrightText: Bosch Rexroth AG
 *
 * SPDX-License-Identifier: MIT
 */
package datalayer.provider.simple;

import java.time.Instant;

import com.boschrexroth.api.AllowedOperation;
import com.boschrexroth.api.MetadataBuilder;
import com.boschrexroth.api.ProviderNodeCallback;
import com.boschrexroth.api.ProviderNodeEvent;
import com.boschrexroth.api.ProviderNodeEventData;
import com.boschrexroth.api.ReferenceType;
import com.boschrexroth.api.Result;
import com.boschrexroth.api.Variant;

import comm.datalayer.NodeClass;

public class ProviderNodeCallbackTimestamp implements ProviderNodeCallback {

    @Override
    public void onBrowse(ProviderNodeEvent ev) {
        System.out.println("onBrowse:");
        ev.callback(Result.OK, null);
    }

    @Override
    public void onCreate(ProviderNodeEventData ev) {
        System.out.println("onCreate:");
    }

    @Override
    public void onMetadata(ProviderNodeEvent ev) {
        System.out.println("onMetadata: '" + ev.getAddress() + "'");
        int mask = AllowedOperation.createMask(AllowedOperation.READ);
        MetadataBuilder meta = MetadataBuilder.create(mask, "Timestamp Variable", "");
        meta.setNodeClass(NodeClass.Variable);
        meta.addReference(ReferenceType.WRITE, "types/datalayer/timestamp");
        Variant v = meta.build();
        ev.callback(Result.OK, v);
    }

    @Override
    public void onRead(ProviderNodeEventData ev) {
        System.out.println("onRead: '" + ev.getAddress());
        Instant time = Instant.now();
        Variant val = Variant.create();
        val.setInstant(time);
        ev.callback(Result.OK, val);
    }

    @Override
    public void onRemove(ProviderNodeEvent ev) {
        System.out.println("onRemove:");
        ev.callback(Result.OK, null);
    }

    @Override
    public void onWrite(ProviderNodeEventData ev) {
        ev.callback(Result.OK, null);
    }

}

File: samples-kotlin/datalayer.provider.simple/app/src/main/kotlin/datalayer/provider/simple/ProviderNodeCallbackTimestamp.kt
/*
 * SPDX-FileCopyrightText: Bosch Rexroth AG
 *
 * SPDX-License-Identifier: MIT
 */
package datalayer.provider.simple

import com.boschrexroth.api.AllowedOperation
import com.boschrexroth.api.MetadataBuilder
import com.boschrexroth.api.ProviderNodeCallback
import com.boschrexroth.api.ProviderNodeEvent
import com.boschrexroth.api.ProviderNodeEventData
import com.boschrexroth.api.ReferenceType
import com.boschrexroth.api.Result
import com.boschrexroth.api.Variant
import comm.datalayer.NodeClass

/**
 * Provider callback for a timestamp node.
 *
 * Reads always return the current wall clock time.
 */
class ProviderNodeCallbackTimestamp : ProviderNodeCallback {

    private val log = System.getLogger(ProviderNodeCallbackTimestamp::class.java.name)

    override fun onBrowse(ev: ProviderNodeEvent) {
        log.log(System.Logger.Level.INFO, "onBrowse:")
        ev.callback(Result.OK, null)
    }

    override fun onCreate(ev: ProviderNodeEventData) {
        log.log(System.Logger.Level.INFO, "onCreate:")
    }

    override fun onMetadata(ev: ProviderNodeEvent) {
        log.log(System.Logger.Level.INFO, "onMetadata: '${ev.address}'")
        val mask = AllowedOperation.createMask(AllowedOperation.READ)
        val meta = MetadataBuilder.create(mask, "Timestamp Variable", "")
        meta.setNodeClass(NodeClass.Variable)
        meta.addReference(ReferenceType.WRITE, "types/datalayer/timestamp")
        ev.callback(Result.OK, meta.build())
    }

    override fun onRead(ev: ProviderNodeEventData) {
        log.log(System.Logger.Level.INFO, "onRead: '${ev.address}'")
        // Snapshot current time on every read request.
        val variant = Variant.create()
        variant.setInstant(ProviderValues.currentTimestampValue())
        ev.callback(Result.OK, variant)
    }

    override fun onRemove(ev: ProviderNodeEvent) {
        log.log(System.Logger.Level.INFO, "onRemove:")
        ev.callback(Result.OK, null)
    }

    override fun onWrite(ev: ProviderNodeEventData) {
        // The sample keeps this node read-only.
        ev.callback(Result.OK, null)
    }
}
